cli/azd/docs/environment-with-hooks.md

```diff
@@ -0,0 +1,172 @@
+
+So, as you are creating a new hook and testing it, make sure to use `azd hooks run <hook name>` if you are reading azd environment. Otherwise, if you try manually running the script, you won't see the azd environment values.
+
+In case you want to reproduce the same example in Windows powershell:
+
+```yaml
+name: demo-hooks
+hooks:
+  preprovision:
+    shell: pwsh
+    run: ./script.ps1
+```
+
+Now let's create `script.ps1` like:
+
+```pwsh
+Write-Output $env:ENV_VAR_FROM_AZD
+
+# just to let you see the output
+sleep 5
+```
+
+### Self environment injection
+
+In case you want to be able to *manually run* the `script` from a hook and without using azd, but you still want to read azd environment.
+
+You might want to allow customers to manually run the scripts in case the hook failed when it was invoked by azd. For example, as a troubleshooting guide, you might be telling folks which script to run to correct missing settings or repair application components. For such cases, consider asking customers to use `azd hooks run <hook name>` as a way to make it simpler for customers and still delegating the environment injection to azd.
+
+But, if you still want to allow the manual invocation, you will need to teach your script how to pull pull the environment values from azd by calling and parsing the output of `azd env get-values`. Below are 2 strategies you can implement on your script.
+
+#### Use script scoped variables
+
+Ideally, the azd environment should be read and used by the script without affecting any other process or the terminal session. See the next code:
+
+```bash
+#!/bin/bash
+
+declare -A azdEnv
+while IFS='=' read -r key value; do
+    value=$(echo "$value" | sed 's/^"//' | sed 's/"$//')
+    azdEnv[$key]=$value
+done <<EOF
+$(azd env get-values)
+EOF
+
+echo "${azdEnv["ENV_VAR_FROM_AZD"]}"
+
+# just to let you see the output
+sleep 5
+```
+
+By using `declare -A azdEnv`, you can allocate a set to write the keys and values from the `azd env get-values` output. Then you can access the values by referencing the environment keys with `"${azdEnv["ENV_VAR_FROM_AZD"]}"`. This strategy would allow you to test or run your script either manually or using `azd hooks run <hook name>`, and without affecting the terminal variables. 
+
+Here is the powershell equivalent.
+
+```pwsh
+$azdEnv = @{}
+$azdEnvRaw = azd env get-values
+
+foreach ($line in $azdEnvRaw -split "`n") {
+    if ($line -match '^(.*?)=(.*)$') {
+        $key = $matches[1]
+        $value = $matches[2] -replace '^"|"$'
+        $azdEnv[$key] = $value
+    }
+}
+
+Write-Output $azdEnv["ENV_VAR_FROM_AZD"]
+
+# just to let you see the output
+Start-Sleep -Seconds 5
+```
```

Comment:
I feel like we should be focusing on the automatic behavior of hooks scripts when run in the context of `azd` and use auto-injection as a standard and less about how to manually load azd env into the scripts since we already do that.

I'd also like to see examples of how we could make downstream calls from scripts... example - calling python script that will automatically make azd env vars available since they were injected into the script environment.

Reply:
100% Agree.   
I'm trying to make it explain and recommend the use of `azd hooks run ..` for testing and how it is easier to let azd to inject the environment.

There are many templates out there, however, using a load-azd-env script.  Some of them exists from before `azd hooks run` was available and because folks wanted to test the script whit azd env.

And yes, I am planning to add examples about the downstream calls for python :)

Reply:
Any downstream calls in a sub process will automatically inherit any environment variables already set in the script including azd .env values.

Reply:
It just feels like all this makes it more confusing when we should just be instructing users to call `azd hooks run <hook>` and it will do the right thing.

Reply:
> It just feels like all this makes it more confusing when we should just be instructing users to call azd hooks run <hook> and it will do the right thing.

I've been asked a few times, after telling folks to use `azd hooks run ...`:  - but what if I want to manually run it?  (usually for more complex scenarios, like a main script invoking other scripts, where only one of the sub-scripts requires azd-env.

I guess we can elevate the `azd hooks run ...` to the learn.microsoft official docs, while leaving this doc as a more advanced/complex guide

Reply:
If they really want to run their hooks independently and not use the feature available, then they'll need ensure they set all the env vars before invoking.  This should be an exception to the run though in my opinion.  Agreed it should be treated more like a footnote in the docs for more complex use cases and not promoted as "the way".

Reply:
Often, my hooks run multiple scripts, but developers only want to run a certain script at a later point. Is there a way to use azd hooks run with a single script?

Reply:
Maybe this is a separate topic (since I'm writing a different doc), but a common scenario or question I see is developers trying to use the outputs of the main.bicep in their language framework, such as Azure service endpoints. For example, an azd script creates a storage account and an OpenAI account and writes those endpoints out to the azd env file. Next, a node or python app wants to discover and use those endpoints to connect via code. Those languages wouldn't really rely on the azd env file - they have their own env files used by framework packages, or rely on the system/container environment variables. Is there a recommended script somewhere for copying azd env values to a separate .env file for a language framework, or one that writes them out to the system for "permanent" use by the app? 

I usually see a hook script used to copy the azd env contents to another location at the end of the provisioning process, but the scenarios in this doc assume the end goal is working with azd environment variables within hook scripts - not outside of them in a language framework/app.

Reply:
@alexwolfmsft , a common strategy for what you described is to use VSCode tasks settings and the Azure Developer CLI VSCode extension to define a `task` for launching your application locally after [injecting the azd-env](https://github.com/Azure-Samples/todo-python-mongo/blob/main/.vscode/tasks.json#L5).  This is the strategy used by the `todo-templates` for running services locally.  As you can see, it depends on VSCode and azd-extension.

When no VSCode or when you don't want to define a task.json file, you can create your own script that start your service locally. The script can call `azd env list -o json` and parse the output to discovered the default azd environment  to use. Then read and use the values to invoke the service. (Basically manually implementing what VSCode does with tasks.json).  I think @pamelafox uses this strategy.

And the 100% manual alternative is where folks needs to cd into the service's folder they need to run and figure what env vars to set before manually launching/starting the service.  For multiple services, folks would use a new console for each service.

Reply:
> Maybe this is a separate topic (since I'm writing a different doc), but a common scenario or question I see is developers trying to use the outputs of the main.bicep in their language framework, such as Azure service endpoints. For example, an azd script creates a storage account and an OpenAI account and writes those endpoints out to the azd env file. Next, a node or python app wants to discover and use those endpoints to connect via code. Those languages wouldn't really rely on the azd env file - they have their own env files used by framework packages, or rely on the system/container environment variables. Is there a recommended script somewhere for copying azd env values to a separate .env file for a language framework, or one that writes them out to the system for "permanent" use by the app?
> 
> I usually see a hook script used to copy the azd env contents to another location at the end of the provisioning process, but the scenarios in this doc assume the end goal is working with azd environment variables within hook scripts - not outside of them in a language framework/app.

@alexwolfmsft - Can you add your scenarios here: https://github.com/Azure/azure-dev/issues/4067

Reply:
Please have a look at this proposal for `azd script run` https://github.com/Azure/azure-dev/pull/4131